pkg/registry/flowcontrol/ensurer/prioritylevelconfiguration.go

```diff
@@ -0,0 +1,199 @@
+		wrapper:  wrapper,
+	}
+}
+
+// NewMandatoryPriorityLevelEnsurer returns a PriorityLevelEnsurer instance that
+// can be used to ensure a set of mandatory PriorityLevelConfiguration configuration objects.
+func NewMandatoryPriorityLevelEnsurer(client flowcontrolclient.PriorityLevelConfigurationInterface) PriorityLevelEnsurer {
+	wrapper := &priorityLevelConfigurationWrapper{
+		client: client,
+	}
+	return &plEnsurer{
+		strategy: newMandatoryEnsureStrategy(wrapper),
+		wrapper:  wrapper,
+	}
+}
+
+// NewPriorityLevelRemover returns a PriorityLevelRemover instance that
+// can be used to remove a set of PriorityLevelConfiguration configuration objects.
+func NewPriorityLevelRemover(client flowcontrolclient.PriorityLevelConfigurationInterface) PriorityLevelRemover {
+	return &plEnsurer{
+		wrapper: &priorityLevelConfigurationWrapper{
+			client: client,
+		},
+	}
+}
+
+// GetPriorityLevelRemoveCandidate returns a list of PriorityLevelConfiguration
+// names that are candidates for removal from the cluster.
+// bootstrap: a set of hard coded PriorityLevelConfiguration configuration
+// objects kube-apiserver maintains in-memory.
+func GetPriorityLevelRemoveCandidate(client flowcontrolclient.PriorityLevelConfigurationInterface, bootstrap []*flowcontrolv1beta1.PriorityLevelConfiguration) ([]string, error) {
+	// TODO(101667): Use a lister here to avoid periodic LIST calls
+	plList, err := client.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list PriorityLevelConfiguration - %w", err)
+	}
+
+	bootstrapNames := sets.String{}
+	for i := range bootstrap {
+		bootstrapNames.Insert(bootstrap[i].GetName())
+	}
+
+	currentObjects := make([]metav1.Object, len(plList.Items))
+	for i := range plList.Items {
+		currentObjects[i] = &plList.Items[i]
+	}
+
+	return getRemoveCandidate(bootstrapNames, currentObjects), nil
+}
+
+type plEnsurer struct {
+	strategy ensureStrategy
+	wrapper  configurationWrapper
+}
+
+func (e *plEnsurer) Ensure(priorityLevels []*flowcontrolv1beta1.PriorityLevelConfiguration) error {
+	for _, priorityLevel := range priorityLevels {
+		if err := ensureConfiguration(e.wrapper, e.strategy, priorityLevel); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *plEnsurer) Remove(priorityLevels []string) error {
+	for _, priorityLevel := range priorityLevels {
+		if err := removeConfiguration(e.wrapper, priorityLevel); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// priorityLevelConfigurationWrapper abstracts all PriorityLevelConfiguration specific logic,
+// with this we can manage all boiler plate code in one place.
+type priorityLevelConfigurationWrapper struct {
+	client flowcontrolclient.PriorityLevelConfigurationInterface
+}
+
+func (fs *priorityLevelConfigurationWrapper) TypeName() string {
+	return "PriorityLevelConfiguration"
+}
+
+func (fs *priorityLevelConfigurationWrapper) Create(object runtime.Object) (runtime.Object, error) {
```

Comment:
What value is this abstraction providing?  Looks like it's just a way to break type safety

Reply:
We want to avoid duplicating the ensure/delete strategy for each type and also for whether the configuration is suggested or mandatory. As a trade off we have the boilerplate code in wrapper.

Reply:
The top level functions accepts typed object, so we provide some sort of type safety. if you look at the main function that drives it, they are all typed.